pkg/synth/acl.go

```diff
@@ -15,153 +15,177 @@ import (
+		collection := generateACLCollectionFromConnection(&s.Defs, &s.Connections[c], opt.SingleACL)
 		collections = append(collections, collection)
 	}
-	collections = append(collections, generateACLCollectionForBlockedSubnets(s, aclSelector))
+	collections = append(collections, generateACLCollectionForBlockedSubnets(s, opt.SingleACL))
 	return ir.MergeACLCollections(collections...)
 }
 
-func generateACLCollectionFromConnection(s *ir.Spec, conn *ir.Connection,
-	aclSelector func(target *ipblock.IPBlock) string) *ir.ACLCollection {
+// (1) check that both resources are supported in nACL generation.
+// (2) check that at least one resource is internal
+// (3) for all pairs of IPAddrs of both resources, rules that allow communication will be created
+func generateACLCollectionFromConnection(s *ir.Definitions, conn *ir.Connection, single bool) *ir.ACLCollection {
+	if !resourceRelevantToACL(conn.Src.Type) {
+		log.Fatalf(fmt.Sprintf(ACLTypeNotSupported, string(conn.Src.Type)))
+	}
+	if !resourceRelevantToACL(conn.Dst.Type) {
+		log.Fatalf(fmt.Sprintf(ACLTypeNotSupported, string(conn.Dst.Type)))
+	}
 	internalSrc := conn.Src.Type != ir.ResourceTypeExternal
 	internalDst := conn.Dst.Type != ir.ResourceTypeExternal
 	internal := internalSrc && internalDst
 	if !internalSrc && !internalDst {
 		log.Fatalf("ACL: Both source and destination are external for connection %v", *conn)
 	}
 	result := ir.NewACLCollection()
-	if !resourceRelevantToACL(conn.Src.Type) && !resourceRelevantToACL(conn.Dst.Type) {
-		return result
-	}
-	var connectionRules []*ir.ACLRule
 	for _, src := range conn.Src.IPAddrs {
+		srcCidrs, src := adjustResource(s, src, conn.Src)
 		for _, dst := range conn.Dst.IPAddrs {
+			dstCidrs, dst := adjustResource(s, dst, conn.Dst)
 			if src == dst && conn.Src.Type != ir.ResourceTypeCidr && conn.Dst.Type != ir.ResourceTypeCidr {
 				continue
 			}
 			for _, trackedProtocol := range conn.TrackedProtocols {
 				reason := explanation{internal: internal, connectionOrigin: conn.Origin, protocolOrigin: trackedProtocol.Origin}
-				protocolRules := allowDirectedConnection(s, src, dst, conn, internalSrc, internalDst, trackedProtocol.Protocol, reason)
-				connectionRules = append(connectionRules, protocolRules...)
+				allowDirectedConnection(src, dst, srcCidrs, dstCidrs, internalSrc, internalDst, trackedProtocol.Protocol, reason, result, single)
 			}
 		}
 	}
-	for _, rule := range connectionRules {
-		acl := result.LookupOrCreate(aclSelector(rule.Target()))
-		if internal {
-			acl.AppendInternal(rule)
-		} else {
-			acl.AppendExternal(rule)
-		}
-	}
 	return result
 }
 
-func allowDirectedConnection(s *ir.Spec, srcCidr, dstCidr *ipblock.IPBlock, conn *ir.Connection, internalSrc, internalDst bool,
-	protocol ir.Protocol, reason explanation) []*ir.ACLRule {
+// create relevant rule(s) to allow traffic
+func allowDirectedConnection(srcCidr, dstCidr *ipblock.IPBlock, srcSubnets, dstSubnets []*ir.ConnResource, internalSrc, internalDst bool,
```

Comment:
I suggest this function is split into two functions: `allowConnectionFronSrc` and `allowConnectionToDst`